src/components/color_picker/color_picker.ts

```diff
@@ -232,64 +255,139 @@ export class ColorPicker extends Component<ColorPickerProps, SpreadsheetChildEnv
-    this.props.onColorPicked(color);
-    this.state.currentColor = color;
+    this.state.currentHslaColor = { ...this.state.currentHslaColor, ...newHsl };
+    this.state.customHexColor = hslaToHex(this.state.currentHslaColor);
+  }
+
+  onColorClick(color: Color) {
+    if (color) {
+      this.props.onColorPicked(toHex(color));
+    }
+  }
+
+  resetColor() {
+    this.props.onColorPicked("");
   }
 
   toggleColorPicker() {
     this.state.showGradient = !this.state.showGradient;
   }
 
-  computeCustomColor(ev: MouseEvent) {
-    this.state.isCurrentColorInvalid = false;
-    this.state.currentColor = computeCustomColor(ev);
-  }
+  dragGradientPointer(ev: MouseEvent) {
+    const initialGradientCoordinates = { x: ev.offsetX, y: ev.offsetY };
+    this.setCustomGradient(initialGradientCoordinates);
 
-  hideMagnifier(_ev: MouseEvent) {
-    this.state.style.display = "none";
+    const initialMousePosition = { x: ev.clientX, y: ev.clientY };
+
+    const onMouseMove = (ev: MouseEvent) => {
+      const currentMousePosition = { x: ev.clientX, y: ev.clientY };
+      const deltaX = currentMousePosition.x - initialMousePosition.x;
+      const deltaY = currentMousePosition.y - initialMousePosition.y;
+
+      const currentGradientCoordinates = {
+        x: initialGradientCoordinates.x + deltaX,
+        y: initialGradientCoordinates.y + deltaY,
+      };
+      this.setCustomGradient(currentGradientCoordinates);
+    };
+
+    startDnd(onMouseMove, () => {});
   }
 
-  showMagnifier(_ev: MouseEvent) {
-    this.state.style.display = "block";
+  dragHuePointer(ev: MouseEvent) {
+    const initialX = ev.offsetX;
+    const initialMouseX = ev.clientX;
+    this.setCustomHue(initialX);
+    const onMouseMove = (ev: MouseEvent) => {
+      const currentMouseX = ev.clientX;
+      const deltaX = currentMouseX - initialMouseX;
+      const x = initialX + deltaX;
+      this.setCustomHue(x);
+    };
+    startDnd(onMouseMove, () => {});
   }
 
-  moveMagnifier(ev: MouseEvent) {
-    this.state.style.background = computeCustomColor(ev);
-    const shiftFromCursor = 10;
-    this.state.style.left = `${ev.offsetX + shiftFromCursor}px`;
-    this.state.style.top = `${ev.offsetY + shiftFromCursor}px`;
+  setHexColor() {
+    const val = (this.hexInputRef.el as HTMLInputElement)?.value;
+    this.state.customHexColor = val;
+    this.state.isCurrentColorInvalid = false;
+    if (!isColorValid(val)) {
+      this.state.isCurrentColorInvalid = true;
+    } else {
+      this.state.currentHslaColor = hexToHSLA(val);
+    }
   }
 
-  get magnifyingGlassStyle() {
-    const { display, background, left, top } = this.state.style;
-    return cssPropertiesToCss({
-      display,
-      "background-color": display === "block" ? background : undefined,
-      left,
-      top,
-    });
+  addCustomColor(ev: Event) {
+    if (!isHSLAValid(this.state.currentHslaColor) || !isColorValid(this.state.customHexColor)) {
+      ev.stopPropagation();
```

Comment:
then we can have `.stop` in the template if we stop the event in all cases

Reply:
:facepalm: